File: src/indicators/atr.js
function calculateATR(data, period) {
  if (!Array.isArray(data) || data.length < period + 1) {
    return null;
  }
  const trs = [];
  for (let i = data.length - period; i < data.length; i++) {
    const high = data[i].high;
    const low = data[i].low;
    const prevClose = data[i - 1].close;
    const tr = Math.max(
      high - low,
      Math.abs(high - prevClose),
      Math.abs(low - prevClose)
    );
    trs.push(tr);
  }
  const sum = trs.reduce((acc, v) => acc + v, 0);
  const atr = sum / period;
  return parseFloat(atr.toFixed(2));
}

module.exports = calculateATR;

File: src/indicators/atr.py
from typing import List, Optional

def calculate_atr(data: List[dict], period: int) -> Optional[float]:
    if len(data) < period + 1:
        return None
    trs = []
    for i in range(-period, 0):
        high = data[i]['high']
        low = data[i]['low']
        prev_close = data[i-1]['close']
        tr = max(
            high - low,
            abs(high - prev_close),
            abs(low - prev_close)
        )
        trs.append(tr)
    atr = sum(trs) / period
    return round(atr, 2)
